### DISCUSSION TITLE
GraphQL: Can't read page.single with there is read permission on page

### DISCUSSION THREAD

**Discussion by codevin** [requarks/wiki#3682] [Error / Bug Report]
In the page-component pull request which I submitted, GraphQL query is used to read the Component page. It fails for general user because it tests 'manage.page' or 'delete.page' permissions. 

Shouldn't it also include 'read.page' also? Because if guest can read a page, GraphQL should also be able to read it. 

File: https://github.com/Requarks/wiki/blob/master/server/graph/resolvers/page.js
Function: 

> 
```javascript
  async single (obj, args, context, info) {
      let page = await WIKI.models.pages.getPageFromDb(args.id)
      if (page) {
        if (WIKI.auth.checkAccess(context.req.user, ['manage:pages', 'delete:pages'], {
          path: page.path,
          locale: page.localeCode
        })
```

**Comment by Smankusors** [comment]
Are you sure your global permission and page rules are correct? You cannot just set global permission without page rules.

**Comment by codevin** [comment]
I will check once more. But guest user can view the pages both where
component is defined and used. That means read permission is there, right?

Even otherwise, just looking at logic of the code, why would you check
manage/delete when graphql is just reading?

~V

**Comment by NGPixel** [comment]
The single endpoint does not check for page rules, only permissions, because it's meant to be used by the admin panel, which is why there's no read:pages permission.

Adding the read:pages permission would require the page rules to be checked and some properties to be marked as manage only.

**Comment by codevin** [comment]
Could you revisit the assumption that GraphQL is only used by admin users or for admin purposes only? 

Because page-components depends on some API way to access the source of page. I guess this is important design decision. 

My point is that page-components will enable more interesting functionality within wikis (like introducing widgets like weather, show table data and so on). It won't work without this capability as of now.

**Comment by NGPixel** [comment]
That's not what I said. This specific resource is for admin. Many other GraphQL resources are used by normal users and guests.

What is your use case for querying the page source?

**Comment by codevin** [comment]
Page components feature, that allows users to define vue components in wiki
page, and use them elsewhere in wiki.
See the submitted pull request: https://github.com/Requarks/wiki/pull/1865

**Comment by NGPixel** [comment]
I don't think using pages to store Vue components is the best idea. It would be very inefficient to fetch and store them in this manner.  I would rather have a dedicated table and resolvers. And possibly inject them directly on the page server-side so they load instantly on the client.

**Comment by codevin** [comment]
It is wiki. Give power to users!  If security is taken care of.

You can't possibly consider all possible situations that users may want to
read from.  Also, it is not only for tables. For e.g. we could allow
sharing of styles and use them to style content in other pages.

**Comment by codevin** [comment]
Can we at least have client side plugins, so those who are interested can have this capability?

I will open a feature request in feedback to get some votes.

**Comment by NGPixel** [comment]
I'm all for it, I just don't agree with the suggested implementation.

A feature request where the implementation can be discussed would be best.

**Comment by codevin** [comment]
If a page contains JSON data table, and some markup to select style of
table, isn't it a clean design? (Here the styles (and table renderer) may
come from server config or from another page in which we define a Vue
component. The latter is not a content page. It needs to be a wiki page
because users can then create and reuse the plugins, learn from different
sites, from each other etc.

And one may use a simple alternate template to fetch data, in case someone
wants to data scraping.

**Comment by krise86** [comment]
I am currently facing this issue.
Trying to get the content of a page trough the graph api as I need to read the source or rendered content from an external application.

How can this be done with read only access to a subset of paths trough the graph api?
